core/zoe/zoe.chainmail

```diff
@@ -0,0 +1,165 @@
+   * should be an amount for the issuer in index 0 of the issuers array.
+   * `payments` is an array of the actual payments to be escrowed,
+   * following the rules in the offerDesc. If the offerDesc rule is
+   * 'offerExactly' or 'offerAtMost', then a payment matching the amount is
+   * expected. If the offerDesc rule is 'wantAtLeast' or 'wantExactly'
+   * then the payment will be ignored and should be undefined.
+   */
+  escrow (offerDesc :List(OfferDescElement), payments :List(Payment)) -> (EscrowReceiptAndClaimPayoff);
+
+  /**
+   * Returns the zoeInstance's array of issuers. This is helpful in 
+   * the case that the user has forgotten the order.
+   */
+  getIssuers() -> (List);
+
+}
+
+interface ZoeGoverningContractFacet () {
+
+  /** 
+   * Instruct Zoe to try reallocating for the given offerIds.
+   * Reallocation is a matrix (array of arrays) where the rows are the
+   * quantities to be paid to the player who made the offer at the same
+   * index in the offerIds array. The reallocation will only happen if
+   * 'offer safety' and conservation of rights are true, as enforced by Zoe.
+   */
+  reallocate (offerIds :List, reallocation :List(List(Quantity)));
+
+  /** 
+   * Eject the offer, taking the current allocation for that offer and
+   * creating payments to be returned to the user. The payments can be
+   * claimed by the user by taking the `claimPayment` ERTP payment
+   * received after escrowing, calling `unwrap` and calling the `claimPayoff`
+   * method on the resulting use object. No 'offer safety' checks are
+   * done here because any previous reallocation performed those checks.
+   */
+  complete (offerIds :List);
+
+  /**
+   * Create an empty offer for recordkeeping purposes (Autoswap uses
+   * this to create the liquidity pool). 
+   */
+  escrowEmptyOffer ( ) -> (offerId);
+  
+  /**
+   * Escrow an offer created by the governing contract. Autoswap uses
+   * this to mint liquidity tokens and add them to the rights managed by
+   * the zoeInstance.
+   */
+  escrowOffer (offerDesc :List, offerPayments :List) -> (offerId);
+
+  /** Burn and validate an escrowReceipt received from the user. */
+  burnEscrowReceipt (escrowReceipt :Payment) -> (Quantity);
+
+  ////// The methods below are pure and have no side-effects. ////////
+
+  /**
+   * Create an array of empty quantities per issuer. Note that if the 
+   * mint is not a basic fungible mint, this may be something other than 0. 
+   */
+  makeEmptyQuantities ( ) -> (List(Quantity));
+
+  /** Get the array of issuers */
+  getIssuers ( ) -> (List(Issuer));
+  
+  /** Get the array of assays */
+  getAssays ( ) -> (List(Assay));
+
+  /** Get the array of strategies, the logic from the assays */
+  getStrategies ( ) -> (List(Strategy));
+
+  /** 
+   * Pass in an array of offerIds and get a matrix (array of arrays)
+   * containing the quantities, in the same order as the offerIds array.
+   */
+  getQuantitiesFor (offerIds :List) -> (List(List(Quantity)));
```

Comment:
pure

Reply:
Whenever coherent, you should expand `List` declarations to `List(SomethingMeaningful)`

Reply:
Good point